pkg/agent/proxy/proxier.go

```diff
@@ -218,61 +255,82 @@ func (p *proxier) installServices() {
 			addedLoadBalancerIPs = svcInfo.LoadBalancerIPStrings()
 		}
 		if len(deletedLoadBalancerIPs) > 0 || len(addedLoadBalancerIPs) > 0 {
-			needUpdate = true
+			needUpdateService = true
 		}
 
-		if !needUpdate {
+		// If neither the Service nor Endpoints of the Service need to be updated, we skip.
+		if !needUpdateService && !needUpdateEndpoints {
 			continue
 		}
 
-		if err := p.ofClient.InstallEndpointFlows(svcInfo.OFProtocol, endpointUpdateList, p.isIPv6); err != nil {
-			klog.Errorf("Error when installing Endpoints flows: %v", err)
-			continue
-		}
-		err := p.ofClient.InstallServiceGroup(groupID, svcInfo.StickyMaxAgeSeconds() != 0, endpointUpdateList)
-		if err != nil {
-			klog.Errorf("Error when installing Endpoints groups: %v", err)
-			p.endpointInstalledMap[svcPortName] = nil
-			continue
-		}
-		// Delete previous flow.
-		if needRemoval {
-			if err := p.ofClient.UninstallServiceFlows(pSvcInfo.ClusterIP(), uint16(pSvcInfo.Port()), pSvcInfo.OFProtocol); err != nil {
-				klog.Errorf("Failed to remove flows of Service %v: %v", svcPortName, err)
-			}
-		}
-		if err := p.ofClient.InstallServiceFlows(groupID, svcInfo.ClusterIP(), uint16(svcInfo.Port()), svcInfo.OFProtocol, uint16(svcInfo.StickyMaxAgeSeconds())); err != nil {
-			klog.Errorf("Error when installing Service flows: %v", err)
-			continue
-		}
-		// Install OpenFlow entries for the ingress IPs of LoadBalancer Service.
-		// The LoadBalancer Service should be accessible from Pod, Node and
-		// external host.
-		var toDelete, toAdd []string
-		if needRemoval {
-			toDelete = pSvcInfo.LoadBalancerIPStrings()
-			toAdd = svcInfo.LoadBalancerIPStrings()
+		if pSvcInfo != nil {
+			klog.V(2).Infof("Updating Service %s %s", svcPortName.Name, svcInfo.String())
 		} else {
-			toDelete = deletedLoadBalancerIPs
-			toAdd = addedLoadBalancerIPs
+			klog.V(2).Infof("Installing Service %s %s", svcPortName.Name, svcInfo.String())
 		}
-		for _, ingress := range toDelete {
-			if ingress != "" {
-				// It is safe to access pSvcInfo here. If this is a new Service,
-				// then toDelete will be an empty slice.
-				if err := p.uninstallLoadBalancerServiceFlows(net.ParseIP(ingress), uint16(pSvcInfo.Port()), pSvcInfo.OFProtocol); err != nil {
-					klog.Errorf("Error when removing LoadBalancer Service flows: %v", err)
-					continue
+
+		if needUpdateEndpoints {
+			err := p.ofClient.InstallEndpointFlows(svcInfo.OFProtocol, endpointUpdateList, p.isIPv6)
+			if err != nil {
+				klog.Errorf("Error when installing Endpoints flows: %v", err)
```

Comment:
I want not to request new changes from you for this PR as it has been pending for some time. But later probably good to either add a comment to this func or even add a log (e.g. still recording failure occurred, and log at the end of the func) about failures will covered by resync which will happen 30s later.